R/check_structure.R

```diff
@@ -0,0 +1,185 @@
+#' @title Checks protocol document structure
+#'
+#' @description This function reads the protocol and checks if the document
+#' structure is correct: chunks have head and tail, required headings are
+#' present and in the right order,...
+#' This function is intended for checking if a protocol is ready to be rendered
+#' and published.
+#'
+#' @inheritParams get_path_to_protocol
+#' @param fail Should the function drop an error in case of a problem?
+#' Defaults to `TRUE` in a non-interactive session and `FALSE` in an interactive
+#' session.
+#'
+#' @return A report of all failed checks.
+#'
+#' @importFrom rmarkdown yaml_front_matter
+#' @importFrom utils head tail
+#' @importFrom assertthat assert_that is.flag noNA
+#' @importFrom stringr str_detect
+#'
+#' @export
+#'
+check_structure <- function(protocol_code, fail = !interactive()) {
+  assert_that(str_detect(protocol_code, "^s[fioap]p-\\d{3}-(nl|en)$"))
+  assert_that(is.flag(fail), noNA(fail))
+
+  x <- load_protocolcheck(x = protocol_code)
+  template_name <-
+    gsub(
+      pattern = "(s\\w{1}p)-\\d*-(\\w{2})",
+      replacement = "template_\\1_\\2",
+      protocol_code
+    )
+  path_to_template <-
+    system.file(
+      file.path("rmarkdown", "templates", template_name, "skeleton"),
+      package = "protocolhelper")
+  files_protocol <- list.files(path = x$path)
+  files_template <- list.files(path = path_to_template)
+
+  # check if file(name)s from template are conserved
+  files_template_i <-
+    c(files_template[files_template != "skeleton.Rmd"], "index.Rmd")
+  difffiles <- files_template_i[!files_template_i %in% files_protocol]
+  difffiles <-
+    difffiles[!grepl("^\\d{2}_appendices.Rmd", difffiles)]
+  difffiles <-
+    difffiles[!grepl("^\\d{2}_subprocotols.Rmd", difffiles)]
+  x$add_error(msg = sprintf("the protocol lacks file(s) %s", difffiles))
+
+  protocol_rmds <- files_protocol[grepl(".Rmd$", files_protocol)]
+  for (file in protocol_rmds) {
+    x <- check_file(file, x, files_template, path_to_template)
+  }
+
+  # check if multiple Rmd files with the same chapter number exist
+  chapters <- protocol_rmds[grepl("^\\d{2}", protocol_rmds)]
+  chapter_numbers <- gsub(
+    pattern = "(?<=^\\d{2})(\\w|\\.)+",
+    replacement = "",
+    x = chapters,
+    perl = TRUE)
+  chapter_numbers <- as.numeric(chapter_numbers)
+  x$add_error(msg =
+                sprintf("multiple file names starting with %s",
+                        formatC(
+                          chapter_numbers[duplicated(chapter_numbers)],
+                          width = 2, flag = "0"
+                        )
+                )
+  )
```

Comment:
While testing branch `gha-check-branch-functions`, this piece of code suddenly dropped this error (which wasn't there when I checked this branch last week):
`Error in x$add_error : $ operator is invalid for atomic vectors`
The tested example didn't have any double chapter numbers, so there shouldn't be an error.  I debugged, and in this case `chapter_numbers[duplicated(chapter_numbers)]` returns `numeric(0)` and `formatC(...)` returns `character(0)`.  Admittedly I didn't make the protocolobject fool proof for all kind of strange input as users don't use it directly, so probably it can't deal with this `character(0)`. I supposed we would only use it to add a string whenever an error should be added.

And also, to be honest, I see no reason to test for chapter numbers here: above we already test if the complete file names (including chapter numbers) are the same as the template, so if the template chapter numbers are fine, this test is superfluous.  I think it would be more relevant to use this test as a unittest to test the template in the package itself: what can a user do if this test appears to fail because there is an error in the template?  This is a problem that should be solved in the template, not a protocol related problem, so it should be solved in protocolhelper and thus should be tested here.  An additional advantage: when including in the unittest, the test only has to be done once when developing the code, and not for every new protocol.

Reply:
The use-case for this test is [when a protocol is created from conversion of a docx protocol](https://github.com/inbo/protocolsource/blob/main/CONTRIBUTING.md#from-an-existing-docx-protocol). It is likely that the chapters in the docx file do not correspond 1-to-1 to the template chapters. The function will in that case copy template .Rmd files that are missing. This can result in multiple .Rmd files with the same chapter number as prefix but a different file name. This test should flag that case. Some harmonization will be needed then. 

I will look into it to fix it.

Reply:
Eh, I'm not sure I fully understand it.  Since chapter names including numbers have to be identical to the template to pass the previous test, this previous test will already fail in the case you are describing and give the user more detailed information on the exact chapter numbers/names that have to be used (while the additional test doesn't give information on the correct filename).  So I still don't understand the additional value of the test: it just adds another error to the list in a case that is already covered by other tests.

(Also, the link seems not to give any information on the creation of a protocol from docx, so maybe the link is wrong and that's why I don't understand?)

Reply:
> Checking only for "other" files than the once in the template is not a valid option, because it is not forbidden to add other files (for instance you could add other reference files like packages.bib or child documents, ... )

That's right.  I thought I also tested in `check_structure()` if all files from the protocol are present in the template, but apparently I didn't.  Should I add this test?  As we can name in the error message of this test which file should be removed, it seems much more informative for a user than the error `multiple file names starting with 0X` of the current test, which doesn't reveal at all which of two files with this chapter number should be kept.

Reply:
Yes, sounds like a good idea to add this test.